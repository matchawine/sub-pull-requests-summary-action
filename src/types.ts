export type GithubActionConfig = {
  baseRefNameFilter: string | null
  token: string
  childPullRequestGithubMarkdownTemplate: string
  loneCommitGithubMarkdownTemplate: string
  githubMarkdownTemplate: string
  childPullRequestOutputMarkdownTemplate: string
  loneCommitOutputMarkdownTemplate: string
  outputMarkdownTemplate: string
}
