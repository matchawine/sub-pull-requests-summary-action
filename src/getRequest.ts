import { GraphQLClient, gql } from "graphql-request"

const query = gql`
  query prNode($prId: ID!) {
    node(id: $prId) {
      ... on PullRequest {
        id
        title
        commits(last: 250) {
          nodes {
            commit {
              abbreviatedOid
              message
              associatedPullRequests(
                first: 10
                orderBy: { field: UPDATED_AT, direction: DESC }
              ) {
                nodes {
                  id
                  title
                  baseRefName
                  state
                  number
                }
              }
            }
          }
        }
      }
    }
  }
`

const endpoint = "https://api.github.com/graphql"

// @ts-ignore
export const executeRequest = async ({ token, prId }) => {
  const variables = { prId }
  const graphQLClient = new GraphQLClient(endpoint, {
    headers: {
      authorization: `Bearer ${token}`,
    },
  })
  return await graphQLClient.request(query, variables)
}
