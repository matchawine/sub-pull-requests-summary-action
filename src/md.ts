import _ from "lodash"
import { GithubActionConfig } from "./types"

type Input = {
  associatedPullRequests: any[]
  loneCommits: any[]
}

_.templateSettings.interpolate = /{{([\s\S]+?)}}/g

type Config = {
  pullRequestMarkdownTemplate: string
  loneCommitMarkdownTemplate: string
  markdownTemplate: string
}

const getMD = (
  { associatedPullRequests, loneCommits }: Input,
  {
    pullRequestMarkdownTemplate,
    loneCommitMarkdownTemplate,
    markdownTemplate,
  }: Config,
) => {
  const getPRText = _.template(pullRequestMarkdownTemplate)
  const getCommitText = _.template(loneCommitMarkdownTemplate)

  const childrenPullRequestMarkdowns = associatedPullRequests.map(getPRText)
  const loneCommitMarkdowns = loneCommits.map(getCommitText)

  return _.template(markdownTemplate)({
    childrenPullRequestMarkdowns,
    loneCommitMarkdowns,
  })
}

export const getGithubMD = (input: Input, config: GithubActionConfig) =>
  getMD(input, {
    pullRequestMarkdownTemplate: config.childPullRequestGithubMarkdownTemplate,
    loneCommitMarkdownTemplate: config.loneCommitGithubMarkdownTemplate,
    markdownTemplate: config.githubMarkdownTemplate,
  })

export const getOutputMD = (input: Input, config: GithubActionConfig) =>
  getMD(input, {
    pullRequestMarkdownTemplate: config.childPullRequestOutputMarkdownTemplate,
    loneCommitMarkdownTemplate: config.loneCommitOutputMarkdownTemplate,
    markdownTemplate: config.outputMarkdownTemplate,
  })
