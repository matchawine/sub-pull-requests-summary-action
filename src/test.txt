sub PR 1
sub PR 2
lone commit 1