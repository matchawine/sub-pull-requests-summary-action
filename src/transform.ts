import _ from "lodash"

// @ts-ignore
export const transform = githubResponse => {
  const pullRequest = githubResponse.node
  const commits = pullRequest.commits.nodes
  const associatedPullRequests = _(commits)
    .map("commit")
    .map("associatedPullRequests")
    .map("nodes")
    .flatten()
    .filter({ baseRefName: "develop", state: "MERGED" })
    .uniqBy("id")
    .value()
  return associatedPullRequests
}
