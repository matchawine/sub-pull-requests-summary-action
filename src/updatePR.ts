import { GraphQLClient, gql } from "graphql-request"
import { GithubActionConfig } from "./types"

const query = gql`
  mutation updateBody($prId: ID!, $body: String!) {
    updatePullRequest(input: { pullRequestId: $prId, body: $body }) {
      pullRequest {
        title
        body
      }
    }
  }
`

const endpoint = "https://api.github.com/graphql"

// @ts-ignore
export const updatePR = async ({ prId, body }, config: GithubActionConfig) => {
  const variables = { prId, body }
  const graphQLClient = new GraphQLClient(endpoint, {
    headers: {
      authorization: `Bearer ${config.token}`,
    },
  })
  return await graphQLClient.request(query, variables)
}
