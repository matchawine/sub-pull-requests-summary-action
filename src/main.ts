import * as core from "@actions/core"
import { executeRequest } from "./getRequest"
import { transform } from "./transform"
import { getMD } from "./md"
import { updatePR } from "./updatePR"
import * as github from "@actions/github"
import { GithubActionConfig } from "./types"
import { config } from "dotenv"

const getConfig = (): GithubActionConfig => {
  const token: string = core.getInput("GITHUB_TOKEN")
  const pullRequestDescriptionTemplate: string = core.getInput(
    "pull-request-description-template",
  )
  const baseRefNameFilter: string | null =
    core.getInput("base-ref-name-filter") || null

  return {
    token,
    baseRefNameFilter,
    pullRequestDescriptionTemplate,
  }
}

async function run(): Promise<void> {
  try {
    const config: GithubActionConfig = getConfig()

    const prId: string = github.context.payload.pull_request?.node_id

    if (!prId)
      throw new Error(
        "No pull request id found. Are you in a pull request workflow?",
      )

    const res = await executeRequest({ prId }, config)
    const v = transform(res, config)
    const md = getMD(v, config)
    console.log("Udpate PR", md)
    updatePR({ prId, body: md }, config)
  } catch (error) {
    if (error instanceof Error) core.setFailed(error.message)
  }
}

run()
