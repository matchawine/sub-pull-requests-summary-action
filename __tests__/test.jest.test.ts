import "dotenv/config"
import { executeRequest } from "../src/getRequest"
import { transform } from "../src/transform"
import { getMD } from "../src/md"

const token = process.env.GITHUB_TOKEN
const prId = "PR_kwDOGzILZc4yIyxe"

describe("Ship description", () => {
  test("Response", async () => {
    const res = await executeRequest({ token, prId })
    expect(res).toMatchSnapshot()
  })

  test("Transform", async () => {
    const res = await executeRequest({ token, prId })
    const transformed = transform(res)
    expect(transformed).toMatchSnapshot()
  })

  test("MD", async () => {
    const res = await executeRequest({ token, prId })
    const transformed = transform(res)
    const md = getMD(transformed)
    expect(md).toMatchSnapshot()
  })
})
