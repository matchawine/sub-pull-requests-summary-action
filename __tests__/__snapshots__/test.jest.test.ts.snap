// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ship description Github MD 1`] = `
"- #18 @DiesIrae
- #19 @DiesIrae

Lone commits:
- [6f890b6](https://github.com/matchawine/sub-pull-requests-summary-action/commit/6f890b6e570c66d8cb28c261528e7e06370f6260) Lone commit 1 @DiesIrae
- [6542788](https://github.com/matchawine/sub-pull-requests-summary-action/commit/6542788e8f830800a21e7a8a865f3f6ef0dbc2ea) Lone commits @DiesIrae"
`;

exports[`Ship description Output MD 1`] = `
"- [#18 Sub PR 1](https://github.com/matchawine/sub-pull-requests-summary-action/pull/18) @DiesIrae
- [#19 Sub PR 2](https://github.com/matchawine/sub-pull-requests-summary-action/pull/19) @DiesIrae"
`;

exports[`Ship description Response 1`] = `
Object {
  "node": Object {
    "commits": Object {
      "nodes": Array [
        Object {
          "commit": Object {
            "abbreviatedOid": "805ed82",
            "associatedPullRequests": Object {
              "nodes": Array [
                Object {
                  "assignees": Object {
                    "nodes": Array [],
                  },
                  "author": Object {
                    "login": "DiesIrae",
                  },
                  "baseRefName": "target-branch",
                  "id": "PR_kwDOGzILZc4zL_V6",
                  "number": 20,
                  "state": "OPEN",
                  "title": "Test Parent PR",
                  "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/20",
                },
                Object {
                  "assignees": Object {
                    "nodes": Array [],
                  },
                  "author": Object {
                    "login": "DiesIrae",
                  },
                  "baseRefName": "parent-pr-branch",
                  "id": "PR_kwDOGzILZc4zL_MY",
                  "number": 18,
                  "state": "MERGED",
                  "title": "Sub PR 1",
                  "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/18",
                },
              ],
            },
            "author": Object {
              "user": Object {
                "login": "DiesIrae",
              },
            },
            "id": "C_kwDOGzILZdoAKDgwNWVkODIzMGRhNzhkNjc2ZWQ5NTEyNTAzNTZkZjFhNjU2NjM1ODQ",
            "message": "Sub PR 1 commit",
            "url": "https://github.com/matchawine/sub-pull-requests-summary-action/commit/805ed8230da78d676ed951250356df1a65663584",
          },
        },
        Object {
          "commit": Object {
            "abbreviatedOid": "05d8493",
            "associatedPullRequests": Object {
              "nodes": Array [
                Object {
                  "assignees": Object {
                    "nodes": Array [],
                  },
                  "author": Object {
                    "login": "DiesIrae",
                  },
                  "baseRefName": "target-branch",
                  "id": "PR_kwDOGzILZc4zL_V6",
                  "number": 20,
                  "state": "OPEN",
                  "title": "Test Parent PR",
                  "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/20",
                },
                Object {
                  "assignees": Object {
                    "nodes": Array [],
                  },
                  "author": Object {
                    "login": "DiesIrae",
                  },
                  "baseRefName": "parent-pr-branch",
                  "id": "PR_kwDOGzILZc4zL_MY",
                  "number": 18,
                  "state": "MERGED",
                  "title": "Sub PR 1",
                  "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/18",
                },
              ],
            },
            "author": Object {
              "user": Object {
                "login": "DiesIrae",
              },
            },
            "id": "C_kwDOGzILZdoAKDA1ZDg0OTNkNWY4YzA0YzVjYThlZjM0NGQ4N2M0ZjgwMTgzYWQ5Y2I",
            "message": "Merge pull request #18 from matchawine/sub-pr1-branch",
            "url": "https://github.com/matchawine/sub-pull-requests-summary-action/commit/05d8493d5f8c04c5ca8ef344d87c4f80183ad9cb",
          },
        },
        Object {
          "commit": Object {
            "abbreviatedOid": "b4cd0f5",
            "associatedPullRequests": Object {
              "nodes": Array [
                Object {
                  "assignees": Object {
                    "nodes": Array [],
                  },
                  "author": Object {
                    "login": "DiesIrae",
                  },
                  "baseRefName": "target-branch",
                  "id": "PR_kwDOGzILZc4zL_V6",
                  "number": 20,
                  "state": "OPEN",
                  "title": "Test Parent PR",
                  "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/20",
                },
                Object {
                  "assignees": Object {
                    "nodes": Array [],
                  },
                  "author": Object {
                    "login": "DiesIrae",
                  },
                  "baseRefName": "parent-pr-branch",
                  "id": "PR_kwDOGzILZc4zL_SN",
                  "number": 19,
                  "state": "MERGED",
                  "title": "Sub PR 2",
                  "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/19",
                },
              ],
            },
            "author": Object {
              "user": Object {
                "login": "DiesIrae",
              },
            },
            "id": "C_kwDOGzILZdoAKGI0Y2QwZjVkOWQyMTcxOGE4YThlZjM3Nzg0ZjM4NzZkZGM4MWI5MDY",
            "message": "Sub PR 2 commit",
            "url": "https://github.com/matchawine/sub-pull-requests-summary-action/commit/b4cd0f5d9d21718a8a8ef37784f3876ddc81b906",
          },
        },
        Object {
          "commit": Object {
            "abbreviatedOid": "43d39b4",
            "associatedPullRequests": Object {
              "nodes": Array [
                Object {
                  "assignees": Object {
                    "nodes": Array [],
                  },
                  "author": Object {
                    "login": "DiesIrae",
                  },
                  "baseRefName": "target-branch",
                  "id": "PR_kwDOGzILZc4zL_V6",
                  "number": 20,
                  "state": "OPEN",
                  "title": "Test Parent PR",
                  "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/20",
                },
                Object {
                  "assignees": Object {
                    "nodes": Array [],
                  },
                  "author": Object {
                    "login": "DiesIrae",
                  },
                  "baseRefName": "parent-pr-branch",
                  "id": "PR_kwDOGzILZc4zL_SN",
                  "number": 19,
                  "state": "MERGED",
                  "title": "Sub PR 2",
                  "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/19",
                },
              ],
            },
            "author": Object {
              "user": Object {
                "login": "DiesIrae",
              },
            },
            "id": "C_kwDOGzILZdoAKDQzZDM5YjQzOGJhNjlmOGY0ZTM4NTg4YzZhMGY1NjdlMjVjYmE2MDA",
            "message": "Merge pull request #19 from matchawine/sub-pr2-branch",
            "url": "https://github.com/matchawine/sub-pull-requests-summary-action/commit/43d39b438ba69f8f4e38588c6a0f567e25cba600",
          },
        },
        Object {
          "commit": Object {
            "abbreviatedOid": "6f890b6",
            "associatedPullRequests": Object {
              "nodes": Array [
                Object {
                  "assignees": Object {
                    "nodes": Array [],
                  },
                  "author": Object {
                    "login": "DiesIrae",
                  },
                  "baseRefName": "target-branch",
                  "id": "PR_kwDOGzILZc4zL_V6",
                  "number": 20,
                  "state": "OPEN",
                  "title": "Test Parent PR",
                  "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/20",
                },
              ],
            },
            "author": Object {
              "user": Object {
                "login": "DiesIrae",
              },
            },
            "id": "C_kwDOGzILZdoAKDZmODkwYjZlNTcwYzY2ZDhjYjI4YzI2MTUyOGU3ZTA2MzcwZjYyNjA",
            "message": "Lone commit 1",
            "url": "https://github.com/matchawine/sub-pull-requests-summary-action/commit/6f890b6e570c66d8cb28c261528e7e06370f6260",
          },
        },
        Object {
          "commit": Object {
            "abbreviatedOid": "6542788",
            "associatedPullRequests": Object {
              "nodes": Array [
                Object {
                  "assignees": Object {
                    "nodes": Array [],
                  },
                  "author": Object {
                    "login": "DiesIrae",
                  },
                  "baseRefName": "target-branch",
                  "id": "PR_kwDOGzILZc4zL_V6",
                  "number": 20,
                  "state": "OPEN",
                  "title": "Test Parent PR",
                  "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/20",
                },
              ],
            },
            "author": Object {
              "user": Object {
                "login": "DiesIrae",
              },
            },
            "id": "C_kwDOGzILZdoAKDY1NDI3ODhlOGY4MzA4MDBhMjFlN2E4YTg2NWYzZjZlZjBkYmMyZWE",
            "message": "Lone commits",
            "url": "https://github.com/matchawine/sub-pull-requests-summary-action/commit/6542788e8f830800a21e7a8a865f3f6ef0dbc2ea",
          },
        },
      ],
    },
    "id": "PR_kwDOGzILZc4zL_V6",
    "title": "Test Parent PR",
  },
}
`;

exports[`Ship description Transform 1`] = `
Object {
  "associatedPullRequests": Array [
    Object {
      "assignees": Object {
        "nodes": Array [],
      },
      "author": Object {
        "login": "DiesIrae",
      },
      "baseRefName": "parent-pr-branch",
      "id": "PR_kwDOGzILZc4zL_MY",
      "number": 18,
      "state": "MERGED",
      "title": "Sub PR 1",
      "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/18",
    },
    Object {
      "assignees": Object {
        "nodes": Array [],
      },
      "author": Object {
        "login": "DiesIrae",
      },
      "baseRefName": "parent-pr-branch",
      "id": "PR_kwDOGzILZc4zL_SN",
      "number": 19,
      "state": "MERGED",
      "title": "Sub PR 2",
      "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/19",
    },
  ],
  "loneCommits": Array [
    Object {
      "abbreviatedOid": "6f890b6",
      "associatedPullRequests": Object {
        "nodes": Array [
          Object {
            "assignees": Object {
              "nodes": Array [],
            },
            "author": Object {
              "login": "DiesIrae",
            },
            "baseRefName": "target-branch",
            "id": "PR_kwDOGzILZc4zL_V6",
            "number": 20,
            "state": "OPEN",
            "title": "Test Parent PR",
            "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/20",
          },
        ],
      },
      "author": Object {
        "user": Object {
          "login": "DiesIrae",
        },
      },
      "id": "C_kwDOGzILZdoAKDZmODkwYjZlNTcwYzY2ZDhjYjI4YzI2MTUyOGU3ZTA2MzcwZjYyNjA",
      "message": "Lone commit 1",
      "url": "https://github.com/matchawine/sub-pull-requests-summary-action/commit/6f890b6e570c66d8cb28c261528e7e06370f6260",
    },
    Object {
      "abbreviatedOid": "6542788",
      "associatedPullRequests": Object {
        "nodes": Array [
          Object {
            "assignees": Object {
              "nodes": Array [],
            },
            "author": Object {
              "login": "DiesIrae",
            },
            "baseRefName": "target-branch",
            "id": "PR_kwDOGzILZc4zL_V6",
            "number": 20,
            "state": "OPEN",
            "title": "Test Parent PR",
            "url": "https://github.com/matchawine/sub-pull-requests-summary-action/pull/20",
          },
        ],
      },
      "author": Object {
        "user": Object {
          "login": "DiesIrae",
        },
      },
      "id": "C_kwDOGzILZdoAKDY1NDI3ODhlOGY4MzA4MDBhMjFlN2E4YTg2NWYzZjZlZjBkYmMyZWE",
      "message": "Lone commits",
      "url": "https://github.com/matchawine/sub-pull-requests-summary-action/commit/6542788e8f830800a21e7a8a865f3f6ef0dbc2ea",
    },
  ],
}
`;
